File: 10/day10.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <iostream>

#include "../utils/utils.h"

using namespace std;

vector<string> make_entries(string line)
{
  vector<string> entry ( line.size() );
  for (int i = 0; i < line.size(); ++i) {
    entry[i] = line[i];
  }

  return entry;
}

int main()
{
  vector<string> lines = open_file("10/day10.txt");
  vector<vector<string> > entries;
  transform(lines.begin(), lines.end(), back_inserter(entries), &make_entries);

  unordered_set<string> openings ({ "(", "<", "[", "{" });
  unordered_map<string, string> closing_map ({ { "(", ")" } , { "<", ">" } , { "[", "]" } , { "{", "}" } });
  vector<vector<string> > errors;
  for (int i = 0; i < entries.size(); ++i) {
    vector<string> stack;
    bool incomplete = true;
    for (int j = 0; j < entries[i].size(); ++j) {
      string c = entries[i][j];
      if (openings.find(c) != openings.end()) {
        stack.push_back(c);
      } else {
        string opening = stack.back();
        stack.pop_back();
        string closing = closing_map[opening];
        if (closing != c) {
          incomplete = false;
          break;
        }
      }
    }
    if (incomplete) {
      vector<string> closings ( stack.size() );
      for (int k = 0; k < stack.size(); ++k) {
        closings[k] = closing_map[stack[k]];
      }
      reverse(closings.begin(), closings.end());
      errors.push_back(closings);
    }
  }

  vector<size_t> scores ( errors.size() );
  for (int i = 0; i < errors.size(); ++i) {
    size_t score = 0;
    for (int j = 0; j < errors[i].size(); ++j) {
      score *= 5;
      if (errors[i][j] == ")") {
        score += 1;
      } else if (errors[i][j] == "]") {
        score += 2;
      } else if (errors[i][j] == "}") {
        score += 3;
      } else {
        score += 4;
      }
    }
    scores[i] = score;
  }

  sort(scores.begin(), scores.end());

  int idx = floor(scores.size() / 2);
  cout << scores[idx] << endl;
  return 0;
}

File: python/day10.py
with open('10/day10.txt', 'r') as f:
  lines = f.readlines()

entries = [list(line.strip()) for line in lines]
openings = set(['(', '<', '[', '{'])
closing_map = {'(': ')', '<': '>', '[': ']', '{': '}'}
errors = []
for entry in entries:
  stack = []
  incomplete = True
  for char in entry:
    if char in openings:
      stack.append(char)
    else:
      opening = stack.pop()
      if closing_map[opening] != char:
        incomplete = False
        break
  if incomplete:
    errors.append([closing_map[opening] for opening in stack][::-1])

scores = []
for err in errors:
  score = 0
  for close in err:
    score *= 5
    if close == ')': score += 1
    elif close == ']': score += 2
    elif close == '}': score += 3
    else: score += 4
  scores.append(score)

scores.sort()
print(scores[len(scores) // 2])

# Part 1
# ans = 0
# for err in errors:
#   if err == ')': ans += 3
#   elif err == ']': ans += 57
#   elif err == '}': ans += 1197
#   else: ans += 25137

